docs/motivation.md

```diff
@@ -1,65 +1,96 @@
 # Motivation
 
-Why do we need another Haskell infrastructure for Nix?
+`haskell.nix` is an alternative to:
+- `Stack` or `cabal-install` build tools
+- [Nixpkgs Haskell infrastructure for Nix](https://haskell4nix.readthedocs.io/)
 
-Doesn't nixpkgs
-provide a sufficiently good Haskell infrastructure already?
+Using Nix instead of Stack or cabal-install gives us all the niceties of Nix:
+deterministic and hermetic builds, caching... Nix allows us to be very precise
+about the GHC compiler to use which only Stack allows to some extent.
```

Comment:
We can't really replace stack or cabal-install, and that's also not the goal.

If you have a cabal or stack project, the goal is to be able to turn this into a reproducible build plan, and use nix to build it (thus getting the full transparency of dependencies, and exact pinning those).

Reply:
We do, stack and cabal-install provide the package-set (aka build plan), which we build with nix and Cabal. In some sense we are in between stack/cabal-install and Cabal (Setup.hs).

But surely one goal is that Haskell.nix does not want to force nix on you :joy: It gives nix super-powers to your cabal-install/stack project (if you want).

Reply:
So iiuc to be more precise:
- we don't use `stack` program at all because we can derive everything from the `stack.yaml` file
- we can use `cabal-install`'s solver to derive a build plan (`plan.json`) but then we don't use `cabal-install` to build/install the packages. Are we actually using `cabal-install` program or `cabal-install` as a library to derive the build plan?
- in theory we don't need neither stack nor cabal-install as we could write our dependencies explicitly ourselves in nix. It's just that it's more convenient to reuse what's already there.

Is that correct? :)

> But surely one goal is that Haskell.nix does not want to force nix on you. It gives nix super-powers to your cabal-install/stack project (if you want).

What does it buy us to keep the compatibility with cabal-install/stack? Could we have purely haskell.nix projects without stack.yaml/cabal.project? What would we lose? (`cabal repl` comes to mind, but is there something else?)

Asking for better understanding and also to document this :)

Reply:
Perhaps it is clearest to say:
- We want to _behave_ identically to a build performed with stack or cabal-install (that's kind of the point!), and we use those tools in order to figure out how to do that.
- But we do not perform the build _itself_ with stack or cabal-install, but rather with nix.

> What does it buy us to keep the compatibility with cabal-install/stack?

AUIU one of haskell.nix's philosophical stances is to get as close as possible to Just Working if the user has a working build with stack/cabal. Sure, you could probably use the pieces of haskell.nix to make something completely independent, but we gain so, so much value from really trying to stick to what the build tool does and letting users configure things in the normal way.

Reply:
> we don't use stack program at all because we can derive everything from the stack.yaml file

I think this is correct but I am not entirely sure.

> Are we actually using cabal-install program or cabal-install as a library to derive the build plan?

We used to run `cabal freeze` (which still generates a build plan) and read the `plan.json` file. Now we use cabal-install as a library as it gives us more details (all the details actually but we are far from using all of them).

> in theory we don't need neither stack nor cabal-install as we could write our dependencies explicitly ourselves in nix. It's just that it's more convenient to reuse what's already there.

With a strong emphasis on **in theory**, yes. You could define your own package set in nix, if you really wanted to.